src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -451,6 +459,25 @@ internal NamespaceOrTypeOrAliasSymbolWithAnnotations BindNamespaceOrTypeOrAliasS
                             if (a.QuestionToken.IsKind(SyntaxKind.QuestionToken))
                             {
                                 type = TypeSymbolWithAnnotations.Create(array, isNullableIfReferenceType: true);
+
+                                if (InExecutableBinder)
+                                {
+                                    // Inside a method body or other executable code, we can pull on NonNullTypes symbol or question IsValueType without causing cycles.
+                                    // Types created outside executable context should be checked by the responsible symbol (the method symbol checks its return type, for instance).
+                                    // We still need to delay that check when binding in an attribute argument
+                                    if (!ShouldCheckConstraintsNullability)
+                                    {
+                                        diagnostics.Add(new LazyMissingNonNullTypesContextDiagnosticInfo(Compilation, NonNullTypesContext, type: default), a.QuestionToken.GetLocation());
+                                    }
+                                    else
```

Comment:
>else [](start = 36, length = 4)

Can or should this also check for IsValueType and IsErrorType as above? 
If so, this block can probably be factored out to a helper method.

Reply:
Right, so the code above seems like it would work (although the extra checks give a known result).
I think the cost of a few wasted instructions is worth the benefit of factored code.

---
In reply to: [217139199](https://github.com/dotnet/roslyn/pull/29809#discussion_r217139199) [](ancestors = 217139199,217129906)